### ISSUE TITLE
Feature: save Message-Id and search email by Message-Id

### ISSUE THREAD

**Issue by etidahouse** [gessnerfl/fake-smtp-server#125]
Hello,
Is it planned in the new release to allow to save the `Message-Id` of the email and thus search an email via the REST API by `Message-Id` ?

Would you be interrested in such PR ?

**Comment by gessnerfl** [comment]
This is not yet planned. The new release will migrate to latest Spring Boot version and Replace the Thymeleaf UI with React. The UI Controllers will be removed so that only REST endpoints will be used in future.

A PR would be great

**Comment by etidahouse** [comment]
PR : https://github.com/gessnerfl/fake-smtp-server/pull/126

**Comment by gessnerfl** [comment]
Hi @etidahouse 
I will include this into the code line but will need more time. Message-ID is not mandatory. Some clients might not support it. In the new version I plan to support search and filter capabilities where message-id can be supported as well.
My idea right now is to first release version 2.x (most probably today) and after integrate your changes to add message ID and then add support for search and filtering on the UI and REST API

**Comment by etidahouse** [comment]
Hello @gessnerfl 

I updated the PR and added a search/filter function

Do you agree with this kind of implementation ?
I can add more tests

**Comment by gessnerfl** [comment]
Hi @etidahouse,
this is fantastic and actually is exactly what I was aiming for. Very close to Lucene syntax.
would be great if you could add more tests e.g. also covering AND/OR expressions.

Already now, thank you for your great contribution.

**Comment by etidahouse** [comment]
Hello @gessnerfl 

I have updated my PR, ready for review

I took the liberty of making a suggestion to improve the development with docker image (to do not install any jdk, npm, node, yarn... version on local machine) and to all use the same dev environment

Do not hesitate to make comments/suggestion

Thank you for this is useful project

**Comment by gessnerfl** [comment]
@etidahouse as mentioned on the PR the syntax for logical operators looks a bit unnatural for me. 
One thing which just came into my mind is if query parsing is really needed. In theory the API could already be defined such as to provide filters as nested objects.

```json
...
"filter": {
    "operator": "OR",
    "expressions": [
        {
            "field": "fromAddress",
            "operator": "EQUALS",
            "value": "sender@example.com"
        },
        {
            "field": "title",
            "operator": "CONTAINS",
            "value": "test"
        }
    ]
}
...
```

What do you think about this idea. For the UI it would be easy to map filters like this. But I would appreciate your feedback a lot and understand your API use case to design the API as generic as possible.

**Comment by etidahouse** [comment]
Hello @gessnerfl 

I have change my PR and implement an new search function

```json
{
  "logicalOperator": "AND"
  "filters": [
    {
      "key": "toAddress",
      "fieldType": "STRING",
      "operator": "EQUAL",
      "value": "address@em.ail"
    },
    {
      "key": "receivedOn",
      "fieldType": "DATE",
      "operator": "LESS_THAN",
      "value": "2023-05-12"
    }
  ],
  "sorts": [
    {
      "key": "receivedOn",
      "direction": "DESC"
    }
  ],
  "page": 0,
  "size": 10,
}
```

This approach is less complex than the previous implementation, but sufficient for the purposes of this project.
The search function is generic and can be applied to other objects. We can easily add a search function for the EmailAttachment or EmailContent object if needed

Can you please review my PR again ?

**Comment by gessnerfl** [comment]
@etidahouse The feature is now available with version 2.1.0.  I adopted the code and the REST interface slightly to simplify some stuff and to adopt it to Java 17 records.
